go/chat/livelocation_test.go

```diff
@@ -0,0 +1,263 @@
+		require.NoError(m.t, err)
+		m.unfurlCh <- []chat1.Coordinate{
+			chat1.Coordinate{
+				Lat: lat,
+				Lon: lon,
+			},
+		}
+	}
+	if !shouldNotify {
+		return
+	}
+	outboxID := storage.GetOutboxIDFromURL(mapurl, convID, msg)
+	mvalid := msg.Valid()
+	mvalid.ClientHeader.OutboxID = &outboxID
+	notMsg := chat1.NewMessageUnboxedWithValid(mvalid)
+	activity := chat1.NewChatActivityWithIncomingMessage(chat1.IncomingMessage{
+		Message: utils.PresentMessageUnboxed(ctx, m.G(), notMsg, uid, convID),
+		ConvID:  convID,
+	})
+	m.G().NotifyRouter.HandleNewChatActivity(ctx, keybase1.UID(uid.String()), chat1.TopicType_CHAT,
+		&activity, chat1.ChatActivitySource_LOCAL)
+}
+
+func checkCoords(t *testing.T, unfurler *mockUnfurler, refcoords []chat1.Coordinate, timeout time.Duration) {
+	select {
+	case coords := <-unfurler.unfurlCh:
+		require.Equal(t, refcoords, coords)
+	case <-time.After(timeout):
+		require.Fail(t, "no map unfurl")
+	}
+}
+
+func updateCoords(livelocation *maps.LiveLocationTracker, coords []chat1.Coordinate,
+	allCoords []chat1.Coordinate) []chat1.Coordinate {
+	for _, c := range coords {
+		livelocation.LocationUpdate(context.TODO(), c)
+		allCoords = append(allCoords, c)
+	}
+	return allCoords
+}
+
+func TestChatSrvLiveLocationCurrent(t *testing.T) {
+	useRemoteMock = false
+	defer func() { useRemoteMock = true }()
+	ctc := makeChatTestContext(t, "TestChatSrvLiveLocationCurrent", 1)
+	defer ctc.cleanup()
+
+	users := ctc.users()
+	tc := ctc.world.Tcs[users[0].Username]
+	chatUI := newMockChatUI()
+	clock := clockwork.NewFakeClock()
+	tc.G.UIRouter = kbtest.NewMockUIRouter(chatUI)
+	timeout := 2 * time.Second
+
+	conv := mustCreateConversationForTest(t, ctc, users[0], chat1.TopicType_CHAT,
+		chat1.ConversationMembersType_IMPTEAMNATIVE)
+
+	unfurler := newMockUnfurler(tc.Context(), t)
+	tc.ChatG.Unfurler = unfurler
+	livelocation := maps.NewLiveLocationTracker(tc.Context())
+	livelocation.SetClock(clock)
+	tc.ChatG.LiveLocationTracker = livelocation
+	tc.ChatG.CommandsSource.(*commands.Source).SetClock(clock)
+
+	mustPostLocalForTest(t, ctc, users[0], conv, chat1.NewMessageBodyWithText(chat1.MessageText{
+		Body: "/location",
+	}))
+	select {
+	case <-chatUI.watchCh:
+	case <-time.After(timeout):
+		require.Fail(t, "no watch position call")
+	}
+
+	coords := []chat1.Coordinate{
+		chat1.Coordinate{
+			Lat: 40.800348,
+			Lon: -73.968784,
+		},
+		chat1.Coordinate{
+			Lat: 40.798688,
+			Lon: -73.973716,
+		},
+		chat1.Coordinate{
+			Lat: 40.795234,
+			Lon: -73.976237,
+		},
+	}
+	updateCoords(livelocation, coords, nil)
+	// no new map yet
+	select {
+	case <-unfurler.unfurlCh:
+		require.Fail(t, "should not have updated yet")
+	default:
+	}
+	clock.Advance(10 * time.Second)
+	checkCoords(t, unfurler, []chat1.Coordinate{coords[2]}, timeout)
+	select {
+	case <-chatUI.clearCh:
+	case <-time.After(timeout):
+		require.Fail(t, "no clear call")
+	}
+}
+
+func TestChatSrvLiveLocation(t *testing.T) {
```

Comment:
can you add a test case with multiple trackers running? maybe be other lingering issues besides calling `close` multiple times

Reply:
I tested this manually, so I am pretty sure it works, but I can add a quick test.

Reply:
oops i misread that close logic